### ISSUE TITLE
Revise and merge development by Team 2 on feature-model layouting

### ISSUE THREAD

**Issue by tthuem** [FeatureIDE/FeatureIDE#649]
![img_20171016_155050](https://user-images.githubusercontent.com/5330357/31618800-3316c03a-b293-11e7-9652-a5c5bc3ea34e.jpg)

**Comment by Kogoro** [comment]
What is not yet implemented?

What is merged into main fork?
- [ ] Avoid that the legend is crossed by connections between features
- [ ] Center constraints again if features have been renamed
- [ ] Left-to-right layouts should use larger circles for alternatives and or groups for better visibility
- [ ] Feature diagrams should scroll automatically if a feature or constraint is moved to the border of the visible area
- [ ] Integrate tree layout library for top-to-bottom and left-to-right feature diagram layouts: [TreeLayout SourceForge](http://treelayout.sourceforge.net/)
- [ ] New notation for dead and false-optional features and redundant constraints similar to problem markers (error, warning, info)
- [ ] Enable to move features using drag and drop in feature order tab (including block selections)

What is ready for merge?
- [ ] Avoid that the legend is crossed by connections between features
- [ ] Center constraints again if features have been renamed
- [ ] Left-to-right layouts should use larger circles for alternatives and or groups for better visibility
- [ ] Feature diagrams should scroll automatically if a feature or constraint is moved to the border of the visible area
- [ ] Integrate tree layout library for top-to-bottom and left-to-right feature diagram layouts: [TreeLayout SourceForge](http://treelayout.sourceforge.net/)
- [ ] New notation for dead and false-optional features and redundant constraints similar to problem markers (error, warning, info)
- [ ] Enable to move features using drag and drop in feature order tab (including block selections)

**Comment by tthuem** [comment]
@Subaro @Kogoro What is merged so far?

**Comment by Subaro** [comment]
@tthuem 
I edited Kogoros list above.

For the features:
* Enable to move features using drag and drop in feature order tab (including block selections)
* Left-to-right layouts should use larger circles for alternatives and or groups for better visibility

I need to check the team2 Fork and see how far they implemented these features.

**Comment by tthuem** [comment]
* Enable to move features using drag and drop in feature order tab (including block selections)

Functionality was implemented and appeared stable. Not sure about code quality though.

* Left-to-right layouts should use larger circles for alternatives and or groups for better visibility

Is something we need to discuss, as it cannot be merged as-is. The large box for circles prevents from selecting many features.

**Comment by tthuem** [comment]
For the record, everything was implemtend in the following fork originally: https://github.com/Henningson/FeatureIDETeam2

**Comment by Subaro** [comment]
The branch is now merged into develop

**Comment by tthuem** [comment]
@Subaro Did you also merge the Wiki and adapted all absolute paths in links? Is the new notation for anomalies also merged? What about the larger circles?

**Comment by tthuem** [comment]
Found numerous issues with the current development branch:

1. Opening a new, but void feature model first layouts all features, but then a new layouting is needed once the analyzer is done with its computations. The reason is that the size of the boxes change.
2. When deactivating the analysis, the new notation for void feature model does not disappear.
3. There are weird error markers for constraints. We agreed that dead features, false-optional features, and void feature models should not be highlighted anymore in constraints.
4. Related to the above: The legend shows something rather confusing for void feature models. The text should be rather "Void feature model".

![a](https://user-images.githubusercontent.com/5330357/32661109-e31aac1c-c625-11e7-94db-a9e45b3c7042.png)

**Comment by tthuem** [comment]
Very likely no further issue than those raised above, but that's rather ugly:

![void](https://user-images.githubusercontent.com/5330357/32927140-4a5d82da-cb4b-11e7-828b-00e0502017b7.png)

**Comment by Subaro** [comment]
@tthuem I added an extra table in Kogoros first post listing all the features that are actually merged. I also created a list for errors and will update them accordingly when I fix them.

**Comment by Subaro** [comment]
@tthuem The Wiki is now merged. All paths should be adapted.

**Comment by Subaro** [comment]
@tthuem 
1) When opening a void model for the first time the starts the analysis and after that the information, that the model is void is cached. But the feature model is loaded before the analysis is finished which leads to the strange behavior that only some feature has an error marker. The cause is that the analysis does not inform the diagram that it finished and therefore the diagram cannot refresh the features.

A question to 2). So when deactivating the automated calculations you want the cached values from an earlier analysis to be erased? Currently he does an analysis and knows that some feature are dead. When deactivating the automated calculation he still has the information from the earlier analysis and the error markers are still there.

To 3) The error marker is not for the feature but to show that the constraint in part of the active explanation. It is clearly redundant to the red border of the constraint and therefore removed.

4) Changed the Text

**Comment by tthuem** [comment]
Besides the update problem of #681, it seems that we are done here.
